sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -3304,6 +3304,26 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
         """.stripMargin).collect()
     }
   }
+
+  test("SPARK-29860: Fix dataType mismatch issue for InSubquery") {
+    withTempView("ta", "tb", "tc", "td") {
+      sql("create temporary view ta as select * from values(cast(1 as Decimal(8, 0))) as ta(id)")
+      sql("create temporary view tb as select * from values(cast(1 as Decimal(9, 0))) as tb(id)")
+      sql("create temporary view tc as select * from values(cast(1 as Decimal(7, 2))), " +
+        "cast(1.23 as Decimal(7,2)) as tc(id)")
+      sql("create temporary view td as select * from values(cast(1 as Decimal(38, 31))) as td(id)")
+      val df1 = sql("select id from ta where id in (select id from tb)")
+      checkAnswer(df1, Array(Row(new java.math.BigDecimal(1))))
```

Comment:
nit: `checkAnswer(df1, Row(new java.math.BigDecimal(1)))`

Reply:
> I think the PR description above should be self-descriptive, so could you please make that more clearer? What's a root cause of this issue, how to fix, brabrabra...

Thanks, updated.

Reply:
thanks, will update it in next commit.